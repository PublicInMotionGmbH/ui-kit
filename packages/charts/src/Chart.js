import React from 'react'
import PropTypes from 'prop-types'
import {
  FlexibleXYPlot, XAxis, YAxis, LineSeries, HorizontalBarSeries, LabelSeries,
  HorizontalGridLines, VerticalBarSeries, VerticalGridLines, Highlight
} from 'react-vis'

import { buildClassName } from '@talixo/shared'

import { generateSeriesClassName } from './utils'

const moduleName = 'chart'

const Components = {
  'bar': VerticalBarSeries,
  'bar-horizontal': HorizontalBarSeries,
  'line': LineSeries
}

const propTypes = {
  /** Additional class name. */
  className: PropTypes.string,

  /** Line chart data.  */
  data: PropTypes.arrayOf(
    PropTypes.shape({
      /** Additional data series className. */
      className: PropTypes.string,

      /** Color of the data series. */
      color: PropTypes.string,

      /** Data information to be displayed in chart. */
      dataItems: PropTypes.arrayOf(PropTypes.shape({
        /** X axis data. */
        x: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,

        /** Y axis data. */
        y: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
      })),

      /** Indicates if data will be shown in the chart. */
      disabled: PropTypes.bool,

      /** Label of data series. */
      label: PropTypes.string
    })
  ),

  /** Hide horizontal grid. */
  hideHorizontalGrid: PropTypes.bool,

  /** Hide vertical grid. */
  hideVerticalGrid: PropTypes.bool,

  /** Should series label be hidden? */
  hideSeriesLabels: PropTypes.bool,

  /** Props which will be applied to react-vis LabelSeries component. */
  seriesLabelProps: PropTypes.object,

  /** Idicates if x axis should be displayed as date */
  timeSeries: PropTypes.bool,

  /** Type of chart (line or bar). */
  type: PropTypes.oneOf(['bar', 'bar-horizontal', 'line']),

  /** X axis title. */
  xAxisTitle: PropTypes.string,

  /** Y axis title. */
  yAxisTitle: PropTypes.string,

  /** Indicates if chart is zoomable, can be applied to chart. */
  zoomable: PropTypes.bool
}

const defaultProps = {
  data: [],
  lineColors: [],
  timeSeries: false,
  type: 'line',
  zoomable: false
}

/**
 * Component which represents Chart.
 *
 * @property {object} [props]
 * @property {string} [props.className]
 * @property {boolean} [props.hideHorizontalGrid]
 * @property {boolean} [props.hideVerticalGrid]
 * @property {boolean} [props.hideSeriesLabels]
 * @property {string} [props.xAxisTitle]
 * @property {string} [props.yAxisTitle]
 * @property {bool} [props.timeSeries]
 * @property {string} [props.type]
 * @property {bool} [props.zoomable]
 *
 * @property {object[]} [props.data]
 * @property {string} [props.data.className]
 * @property {string} [props.data.color]
 * @property {object} [props.data.dataItems]
 * @property {string|number} props.data.dataItems.x
 * @property {string|number} props.data.dataItems.y
 * @property {bool} [props.data.disabled]
 * @property {string} [props.data.label]
 *
 * @property {object} [state]
 * @property {number} [state.bottom]
 * @property {number} [state.left]
 * @property {number} [state.right]
 * @property {number} [state.top]
 *
 * @class
 */
class Chart extends React.Component {
  state = {
    lastDrawLocation: null
  }

  /**
   * Return a string for xType property of FlexibleXYPlot
   *
   * @returns {string}
   */
  getChartType = () => {
    const { timeSeries, type } = this.props
    if (timeSeries) {
      return 'time'
    }
    if (type === 'bar') {
      return 'ordinal'
    }
    return 'linear'
  }

  /**
   * Returns Data Series with its labels.
   *
   * @param item
   * @param index
   * @returns ReactElement[]
   */

  getSeries = (item, index) => {
    if (item.disabled) {
      return null
    }

    const { hideSeriesLabels, dataSeriesProps, seriesLabelProps, type } = this.props
    const RenderComponent = Components[type]
    const isLineChart = type === 'line'

    const itemProps = {
      data: item.dataItems,
      color: item.color,
      className: generateSeriesClassName(index, item.className),
      key: `${item.title || item.label || item.x}-${index}`,
      ...dataSeriesProps
    }

    const series = (
      <RenderComponent
        animate
        style={isLineChart ? {fill: 'none'} : {}}
        {...itemProps}
      />
    )

    const label = !hideSeriesLabels
      ? <LabelSeries data={item.dataItems} key={itemProps.key} {...seriesLabelProps} />
      : null

    return [series, label]
  }

  /**
   * Updates state when area is zoomed
   *
   * @param {object} area
   * @param {number} area.top
   * @param {number} area.right
   * @param {number} area.bottom
   * @param {number} area.left
   */
  onBrushEnd = (area) => {
    this.setState({ lastDrawLocation: area })
  }

  onDrag = (area) => {
    const defaultDrawLocation = { top: 0, bottom: 0, left: 0, right: 0 }
    const { lastDrawLocation } = this.state
    const location = lastDrawLocation || defaultDrawLocation
    this.setState({
      lastDrawLocation: {
        bottom: location.bottom + (area.top - area.bottom),
        left: location.left - (area.right - area.left),
        right: location.right - (area.right - area.left),
        top: location.top + (area.top - area.bottom)
      }
    })
  }

  render () {
    const { lastDrawLocation } = this.state
    const { onBrushEnd, getChartType, getSeries } = this
    const {
      className, data, dataSeriesProps, lineProps, xAxisTitle, yAxisTitle,
      hideHorizontalGrid, hideVerticalGrid, type, zoomable, ...passedProps
    } = this.props

    const wrapperCls = buildClassName(moduleName, className)
    const xType = !passedProps.xType && getChartType()

    return (
      <FlexibleXYPlot
        animation
        xType={xType}
        className={wrapperCls}
        xDomain={lastDrawLocation && [lastDrawLocation.left, lastDrawLocation.right]}
        yDomain={lastDrawLocation && [lastDrawLocation.bottom, lastDrawLocation.top]}
        {...passedProps}
      >
        { !hideHorizontalGrid && <HorizontalGridLines /> }
        { !hideVerticalGrid && <VerticalGridLines /> }
        { data.map((item, index) => getSeries(item, index))}
        {
          zoomable && type === 'line' &&
          <Highlight
            color={null}
            onBrushEnd={onBrushEnd}
            onDrag={this.onDrag}
          />
        }
        <YAxis title={yAxisTitle} />
        <XAxis title={xAxisTitle} />
      </FlexibleXYPlot>
    )
  }
}

Chart.propTypes = propTypes

Chart.defaultProps = defaultProps

Chart.displayName = 'Chart'

export default Chart
