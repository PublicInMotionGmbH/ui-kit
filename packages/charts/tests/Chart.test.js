import React from 'react'
import Chart from '../src/Chart'
import { shallow } from 'enzyme'

const HIGHLIGHT = 'Highlight'
const PLOT = 'FlexibleXYPlot'
const BAR_SERIES = 'VerticalBarSeries'
const LINE_SERIES = 'LineSeries'

const testData = [
  {
    className: 'className',
    color: null,
    title: 'Line 1',
    disabled: false,
    dataItems: [{ x: 1495490400000, y: 5 }, { x: 1496008800000, y: 2 }, { x: 1496181600000, y: 2 }, { x: 1497304800000, y: 2 }]
  },
  {
    className: 'className',
    color: null,
    title: 'Line 2',
    disabled: false,
    dataItems: [{ x: 1495490400000, y: 5 }, { x: 1496008800000, y: 2 }, { x: 1496181600000, y: 2 }, { x: 1497304800000, y: 2 }]
  }
]
const testDataDisabled = [
  {
    className: 'className',
    color: null,
    title: 'Line 1',
    disabled: false,
    dataItems: [{ x: 1495490400000, y: 5 }, { x: 1496008800000, y: 2 }, { x: 1496181600000, y: 2 }, { x: 1497304800000, y: 2 }]
  },
  {
    className: 'className',
    color: null,
    title: 'Line 2',
    disabled: true,
    dataItems: [{ x: 1495490400000, y: 5 }, { x: 1496008800000, y: 2 }, { x: 1496181600000, y: 2 }, { x: 1497304800000, y: 2 }]
  }
]

const createProps = (props = {}) => ({
  data: testData,
  type: 'line',
  zoomable: false,
  ...props
})
const defaultProps = createProps()
const createWrapper = (props = defaultProps) => shallow(<Chart {...props} />)

describe('<Chart>', () => {
  describe('when rendered', () => {
    it('should render properly', () => {
      const wrapper = createWrapper()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when disabled data is passed', () => {
    it('should render only one line', () => {
      const props = createProps({ data: testDataDisabled, type: 'line' })
      const wrapper = createWrapper(props)
      expect(wrapper.find(LINE_SERIES)).toHaveLength(1)
    })
  })

  describe('when zoomable line chart is rendered', () => {
    it('should pass style.fill=none to rendered component', () => {
      const props = createProps({ type: 'line', zoomable: true })
      const wrapper = createWrapper(props)
      expect(wrapper.find(LINE_SERIES).forEach(node => {
        expect(node.props().style.fill).toEqual('none')
      }))
    })

    it('should render highlight component', () => {
      const props = createProps({ type: 'line', zoomable: true })
      const wrapper = createWrapper(props)
      expect(wrapper.find(HIGHLIGHT).exists()).toBe(true)
    })

    it('should set xType prop of FLexibleXYPlot to `linear`', () => {
      const props = createProps({ type: 'line', zoomable: true })
      const wrapper = createWrapper(props)
      expect(wrapper.find(PLOT).props().xType).toBe('linear')
    })

    describe('when xAxis is of type `time`', () => {
      it('should set xType prop of FLexibleXYPlot to `time`', () => {
        const props = createProps({ type: 'line', zoomable: true, timeSeries: true })
        const wrapper = createWrapper(props)
        expect(wrapper.find(PLOT).props().xType).toBe('time')
      })
    })

    describe('when zomm area is selected', () => {
      const area = { top: 0, bottom: 0, left: 1495490400000, right: 1496008800000 }
      const brushArea = { bottom: 2, left: 1495490400000, right: 1496008800000, top: 4 }
      const dragArea = { bottom: 3, left: 1496008800000, right: 1496181600000, top: 4 }

      it('should set state.lastDrawLocation to equal area', () => {
        const props = createProps({ type: 'line', zoomable: true, timeSeries: true })
        const wrapper = createWrapper(props)
        wrapper.find(HIGHLIGHT).props().onBrushEnd(area)
        wrapper.update()
        expect(wrapper.state().lastDrawLocation).toEqual(area)
      })

      it('should set lastDrawLocation to given area', () => {
        const props = createProps({ type: 'line', zoomable: true, timeSeries: true })
        const wrapper = createWrapper(props)
        wrapper.find(HIGHLIGHT).props().onBrushEnd(area)
        wrapper.update()
        expect(wrapper.find(PLOT).props().xDomain).toEqual([area.left, area.right])
      })

      it('should set lastDrawLocation to given area after onDrag event', () => {
        const props = createProps({ type: 'line', zoomable: true, timeSeries: true })
        const wrapper = createWrapper(props)
        wrapper.find(HIGHLIGHT).props().onDrag(dragArea)
        expect(wrapper.state().lastDrawLocation).toEqual(dragArea)
      })

      it('should set lastDrawLocation to given area after onBrushEnd event', () => {
        const props = createProps({ type: 'line', zoomable: true, timeSeries: true })
        const wrapper = createWrapper(props)
        wrapper.find(HIGHLIGHT).props().onBrushEnd(brushArea)
        wrapper.update()
        expect(wrapper.find(PLOT).props().xDomain).toEqual([brushArea.left, brushArea.right])
        expect(wrapper.find(PLOT).props().yDomain).toEqual([brushArea.bottom, brushArea.top])
      })
    })
  })

  describe('when bar chart is rendered', () => {
    it('should pass style.fill=none to rendered component', () => {
      const props = createProps({ type: 'bar', zoomable: true })
      const wrapper = createWrapper(props)
      expect(wrapper.find(BAR_SERIES).forEach(node => {
        expect(node.props().style.fill).not.toEqual('null')
      }))
    })

    it('should set xType prop of FLexibleXYPlot to `ordinal`', () => {
      const props = createProps({ type: 'bar', zoomable: true })
      const wrapper = createWrapper(props)
      expect(wrapper.find(PLOT).props().xType).toBe('ordinal')
    })

    it('should not render Hightlight component', () => {
      const props = createProps({ type: 'bar', zoomable: true })
      const wrapper = createWrapper(props)
      expect(wrapper.find(HIGHLIGHT).exists()).toBe(false)
    })
  })
})
