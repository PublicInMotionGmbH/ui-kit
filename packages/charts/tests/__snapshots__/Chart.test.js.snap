// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart> when rendered should render properly 1`] = `
<FlexibleXYPlot
  animation={true}
  className="talixo-chart"
  lineColors={Array []}
  timeSeries={false}
  xDomain={null}
  xType="linear"
  yDomain={null}
>
  <HorizontalGridLines
    attr="y"
    direction="horizontal"
  />
  <VerticalGridLines
    attr="x"
    direction="vertical"
  />
  <LineSeries
    animate={true}
    className="talixo-chart-series className talixo-chart-series--0"
    color={null}
    curve={null}
    data={
      Array [
        Object {
          "x": 1495490400000,
          "y": 5,
        },
        Object {
          "x": 1496008800000,
          "y": 2,
        },
        Object {
          "x": 1496181600000,
          "y": 2,
        },
        Object {
          "x": 1497304800000,
          "y": 2,
        },
      ]
    }
    getNull={[Function]}
    key="Line 1-0"
    opacity={1}
    stack={false}
    strokeStyle="solid"
    style={
      Object {
        "fill": "none",
      }
    }
  />
  <LabelSeries
    animation={false}
    className=""
    data={
      Array [
        Object {
          "x": 1495490400000,
          "y": 5,
        },
        Object {
          "x": 1496008800000,
          "y": 2,
        },
        Object {
          "x": 1496181600000,
          "y": 2,
        },
        Object {
          "x": 1497304800000,
          "y": 2,
        },
      ]
    }
    getLabel={[Function]}
    key="Line 1-0-label"
    rotation={0}
    stack={false}
    style={Object {}}
  />
  <LineSeries
    animate={true}
    className="talixo-chart-series className talixo-chart-series--1"
    color={null}
    curve={null}
    data={
      Array [
        Object {
          "x": 1495490400000,
          "y": 5,
        },
        Object {
          "x": 1496008800000,
          "y": 2,
        },
        Object {
          "x": 1496181600000,
          "y": 2,
        },
        Object {
          "x": 1497304800000,
          "y": 2,
        },
      ]
    }
    getNull={[Function]}
    key="Line 2-1"
    opacity={1}
    stack={false}
    strokeStyle="solid"
    style={
      Object {
        "fill": "none",
      }
    }
  />
  <LabelSeries
    animation={false}
    className=""
    data={
      Array [
        Object {
          "x": 1495490400000,
          "y": 5,
        },
        Object {
          "x": 1496008800000,
          "y": 2,
        },
        Object {
          "x": 1496181600000,
          "y": 2,
        },
        Object {
          "x": 1497304800000,
          "y": 2,
        },
      ]
    }
    getLabel={[Function]}
    key="Line 2-1-label"
    rotation={0}
    stack={false}
    style={Object {}}
  />
  <YAxis
    attr="y"
    attrAxis="x"
    orientation="left"
  />
  <XAxis
    attr="x"
    attrAxis="y"
    orientation="bottom"
  />
</FlexibleXYPlot>
`;
